## Đồ án 3 - Linear Regression

Họ và tên: Lê Trọng Anh Tú

MSSV: 20127091

Môn học: Toán ứng dụng và thống kê

Lớp: 20CLC07

#### 0. Các thư viện được sử dụng

In [1]:
import numpy as np
import pandas as pd

#### 1. Đọc dữ liệu từ file

Ta chọn file `wine.csv` làm dữ liệu để xây dựng mô hình. Dữ liệu được ngăn cách bởi dấu `';'` nên ta truyền tham số `sep = ';'` vào để hàm đọc dữ liệu tự tách các dữ liệu ra với nhau.

In [2]:
df = pd.read_csv('wine.csv', sep = ';')

Output của dữ liệu vừa đọc được từ `wine.csv`

In [3]:
df

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.4             0.700         0.00             1.9      0.076   
1               7.8             0.880         0.00             2.6      0.098   
2               7.8             0.760         0.04             2.3      0.092   
3              11.2             0.280         0.56             1.9      0.075   
4               7.4             0.700         0.00             1.9      0.076   
...             ...               ...          ...             ...        ...   
1194            7.0             0.745         0.12             1.8      0.114   
1195            6.2             0.430         0.22             1.8      0.078   
1196            7.9             0.580         0.23             2.3      0.076   
1197            7.7             0.570         0.21             1.5      0.069   
1198            7.7             0.260         0.26             2.0      0.052   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    11.0                    34  0.99780  3.51       0.56   
1                    25.0                    67  0.99680  3.20       0.68   
2                    15.0                    54  0.99700  3.26       0.65   
3                    17.0                    60  0.99800  3.16       0.58   
4                    11.0                    34  0.99780  3.51       0.56   
...                   ...                   ...      ...   ...        ...   
1194                 15.0                    64  0.99588  3.22       0.59   
1195                 21.0                    56  0.99633  3.52       0.60   
1196                 23.0                    94  0.99686  3.21       0.58   
1197                  4.0                     9  0.99458  3.16       0.54   
1198                 19.0                    77  0.99510  3.15       0.79   

      alcohol  quality  
0         9.4        5  
1         9.8        5  
2         9.8        5  
3         9.8        6  
4         9.4        5  
...       ...      ...  
1194      9.5        6  
1195      9.5        6  
1196      9.5        6  
1197      9.8        6  
1198     10.9        6  

[1199 rows x 12 columns]

#### 2. Xây dựng mô hình sử dụng toàn bộ 11 thuộc tính được cung cấp 

Ta chuyển dữ liệu từ kiểu `DataFrame` về `Numpy` bằng hàm `to_numpy()`

In [4]:
data_set = df.to_numpy()
data_set

array([[ 7.4 ,  0.7 ,  0.  , ...,  0.56,  9.4 ,  5.  ],
       [ 7.8 ,  0.88,  0.  , ...,  0.68,  9.8 ,  5.  ],
       [ 7.8 ,  0.76,  0.04, ...,  0.65,  9.8 ,  5.  ],
       ...,
       [ 7.9 ,  0.58,  0.23, ...,  0.58,  9.5 ,  6.  ],
       [ 7.7 ,  0.57,  0.21, ...,  0.54,  9.8 ,  6.  ],
       [ 7.7 ,  0.26,  0.26, ...,  0.79, 10.9 ,  6.  ]])

Tìm hệ số $\theta$ của mô hình

In [5]:
def linear_regression(data_set):
    X = data_set[:, :-1] # Lấy tất cả các cột làm biến đầu vào (trừ cột cuối cùng là output)
    y = data_set[:, -1] # Lấy cột cuối cùng làm output
    theta = np.linalg.inv(X.T @ X) @ X.T @ y
    return theta

In [6]:
res_a = linear_regression(data_set)
np.set_printoptions(suppress = True) # Tránh việc in ra các giá trị e-03, e+00...
res_a

array([ 0.00592516, -1.10803754, -0.26304628,  0.01532228, -1.73050274,
        0.00380142, -0.003899  ,  4.33858768, -0.45853548,  0.72971866,
        0.30885865])

Mô hình khi sử dụng toàn bộ 11 thuộc tính được cung cấp có các hệ số $\theta_i$ như sau

In [7]:
op_a = np.array([['x', 'đặc trưng', 'theta', 'hệ số']])
for i in range(data_set.shape[1] - 1):
    tmp = np.array([['x_' + str(i + 1), df.columns[i], 'theta_' + str(i + 1), res_a[i]]])
    op_a = np.append(op_a, tmp, axis = 0)
op_a = pd.DataFrame(data = op_a[1:, :], columns = op_a[0, :])

op_a

x             đặc trưng     theta                  hệ số
0    x_1         fixed acidity   theta_1   0.005925161374105113
1    x_2      volatile acidity   theta_2    -1.1080375422629067
2    x_3           citric acid   theta_3    -0.2630462837012381
3    x_4        residual sugar   theta_4   0.015322283066645292
4    x_5             chlorides   theta_5     -1.730502743057372
5    x_6   free sulfur dioxide   theta_6   0.003801419076862004
6    x_7  total sulfur dioxide   theta_7  -0.003898998694536247
7    x_8               density   theta_8      4.338587684499409
8    x_9                    pH   theta_9   -0.45853547521688753
9   x_10             sulphates  theta_10     0.7297186624705415
10  x_11               alcohol  theta_11    0.30885864844891386

#### 3. Xây dựng mô hình chỉ sử dụng một đặc trưng cho kết quả tốt nhất

Phần này ta sử dụng phương pháp K-Fold Cross Validation để tìm đặc trưng tốt nhất cho việc xây dựng mô hình đạt hiệu quả cao

#### 3.1. Tạo mẫu từ bộ dữ liệu

Ta sử dụng `k = 6` trong đồ án này

In [8]:
k = 6
n = data_set.shape[0]
samples = int(n / k) * np.ones(k)
i = 0
while n - np.sum(samples) != 0:
    samples[i] += 1
    i += 1
print(samples)

[200. 200. 200. 200. 200. 199.]


#### 3.2. Tính giá trị sai lệch theo phương pháp K-Fold Cross Validation

Tính giá trị sai lệch dựa vào đặc trưng ở vị trí `idx` và mẫu test ở vị trí `k`

In [9]:
def cross_validation(idx, k_test):
    data_k = data_set[:, [idx, data_set.shape[1] - 1]]

    start_test = int(np.sum(samples[:k_test]))
    end_test = int(np.sum(samples[:k_test + 1]) - 1)

    test = data_k[start_test:end_test + 1, :]
    training = np.append(data_k[:start_test, :], data_k[end_test + 1:, :], axis = 0)

    theta_train = linear_regression(training)

    loss_array = np.zeros(test.shape[0])
    for i in range(test.shape[0]):
        loss_array[i] = np.abs(test[i, -1] - theta_train @ test[i, :-1].T)

    return np.average(loss_array)

#### 3.3. Tìm đặc trưng tốt nhất

Đặc trưng tốt nhất là đặc trưng có giá trị sai lệch trung bình nhỏ nhất trong tất cả các đặc trưng được cung cấp

In [10]:
# Tìm giá trị sai lệch trung bình cho 11 đặc trưng được cung cấp
calc = np.zeros((data_set.shape[1] - 1, 2))
for i in range(data_set.shape[1] - 1):
    temp = 0
    for j in range(k): temp += cross_validation(i, j)
    calc[i][0] = temp / k
    calc[i][1] = i

# Sort tăng dần theo giá trị sai lệch trung bình
calc = calc[calc[:, 0].argsort()]

# Hiển thị ra màn hình
op_b = np.array([['đặc trưng', 'giá trị sai lệch']])
for i in range(data_set.shape[1] - 1):
    tmp = np.array([[df.columns[int(calc[i][1])], calc[i][0]]])
    op_b = np.append(op_b, tmp, axis = 0)
op_b = pd.DataFrame(data = op_b[1:, :], columns = op_b[0, :])
op_b

đặc trưng    giá trị sai lệch
0                alcohol  0.5490115616701011
1                density  0.7078487050943097
2                     pH  0.7309991988159452
3          fixed acidity   1.054806874175808
4              sulphates    1.08597890510556
5       volatile acidity   1.846487704639521
6         residual sugar  2.0148893668391956
7              chlorides  2.0936369885451094
8            citric acid   2.598740705513529
9    free sulfur dioxide     2.8451016413556
10  total sulfur dioxide  3.1411396884744636

Đặc trưng tốt nhất: **alcohol**

Tìm mô hình chỉ dùng duy nhất 1 đặc trưng là **alcohol**

In [11]:
alcohol_theta = linear_regression(data_set[:, [df.columns.get_loc("alcohol"), data_set.shape[1] - 1]])
print('Mô hình nếu chỉ sử dụng đặc trưng Alcohol: y =', str(alcohol_theta[0]) + 'x')

Mô hình nếu chỉ sử dụng đặc trưng Alcohol: y = 0.543705524123724x


#### 4. Xây dựng mô hình cho kết quả tốt nhất

#### 4.1. Mô hình tuyến tính không có tung độ gốc (y = ax)

In [12]:
X = data_set[:, :-1]
y = data_set[:, -1]
residual = np.linalg.norm(y - X @ res_a)
print("Residual vector value of model 'y = ax' is", residual)

Residual vector value of model 'y = ax' is 22.12434596534916


#### 4.2. Mô hình tuyến tính có tung độ gốc (y = ax + b)

In [13]:
X = data_set[:, :-1]
y = data_set[:, -1]

# Thêm cột toàn số 1
ones_column = np.ones((X.shape[0], 1))
X = np.concatenate((ones_column, X), axis = 1)

# Tìm hệ số thoả mô hình
theta = np.linalg.pinv(X.T @ X) @ X.T @ y

# Tính chuẩn vector phần dư và in ra màn hình
residual = np.linalg.norm(y - X @ theta)
print("Residual vector value of model 'y = ax + b' is", residual)

Residual vector value of model 'y = ax + b' is 22.094716807791656


#### 4.3. Mô hình parabol hoàn chỉnh (y = ax^2 + bx + c)

In [14]:
X = data_set[:, :-1]
y = data_set[:, -1]

# Thêm cột toàn số 1 và cột X^2
ones_column = np.ones((X.shape[0], 1))
x_square = X**2
X = np.concatenate((ones_column, X, x_square), axis = 1)

# Tìm hệ số thoả mô hình
theta = np.linalg.pinv(X.T @ X) @ X.T @ y

# Tính chuẩn vector phần dư và in ra màn hình
residual = np.linalg.norm(y - X @ theta)
print("Residual vector value of model 'y = ax^2 + bx + c' is", residual)

Residual vector value of model 'y = ax^2 + bx + c' is 21.617693192282466


#### 4.4. Kiểm tra xem X có số nào nhỏ hơn 0 hay không (để xác định có lấy ln(X) được hay không)

In [15]:
count = 0
for i in range(X.shape[0]):
    for j in range(X.shape[1]):
        if X[i][j] <= 0: count += 1
print(count)

174


Do có 174 phần tử $X_{i,j}<0$ nên ta không thể lấy $ln(X)$ được 

#### 4.5. Mô hình log - tuyến tính (ln(y) = a + bx)

In [16]:
X = data_set[:, :-1]
y = data_set[:, -1]

# Thêm cột toàn số 1 vào X
ones_column = np.ones((X.shape[0], 1))
X = np.concatenate((ones_column, X), axis = 1)

# Đổi thành ln(y)
ln_y = np.log(y)

# Tìm hệ số thoả mô hình
theta = np.linalg.pinv(X.T @ X) @ X.T @ ln_y

# Tính chuẩn vector phần dư và in ra màn hình
residual = np.linalg.norm(y - np.exp(X @ theta))
print("Residual vector value of model 'ln(y) = a + bx' is", residual)

Residual vector value of model 'ln(y) = a + bx' is 22.12508309025727
